app/models/transformation_mapping.rb

```diff
@@ -6,4 +6,72 @@ class TransformationMapping < ApplicationRecord
   def destination(source)
     transformation_mapping_items.find_by(:source => source).try(:destination)
   end
+
+  def validate_vms(csv_list = nil)
+    csv_list.present? ? identify_vms(csv_list) : select_vms
+  end
+
+  private
+
+  def select_vms
+    # TBD
+  end
+
+  def identify_vms(csv_list)
+    valid_list = []
+    invalid_list = []
+    conflict_list = []
+
+    csv_list.each do |row|
+      vm_name = row['Name']
+
+      conditions = {:name => vm_name}
+
+      # construct conditions or selections based on MIQ exported csv columns
+      conditions[:uid_ems] = row['UID'] if row['UID'].present?
+      vms = Vm.where(conditions)
+      vms = vms.select { |vm| vm.host.name == row['Host'] } if row['Host'].present?
+
+      invalid_list << describe_non_vm(vm_name) if vms.empty?
+
+      if vms.size == 1
+        reason = validate_vm(vms.first)
+        (reason.present? ? invalid_list : valid_list) << describe_vm(vms.first, reason)
+      else
+        vms.each { |vm| conflict_list << describe_vm(vm, "Duplicate") }
+      end
+    end
+
+    {
+      "valid_vms"    => valid_list,
+      "invalid_vms"  => invalid_list,
+      "conflict_vms" => conflict_list
+    }
+  end
+
+  def describe_non_vm(vm_name)
+    {
+      N_("Name")   => vm_name,
+      N_("Reason") => "VM does not exist"
```

Comment:
I don't understand why we are translating keys of a Hash.  Is this for presentation purposes?  If so, then this doesn't belong here at all, as presentation logic does not belong in a model.  Instead, the model should return a "normal" Hash, and something else can translate that for presentation purposes.

Reply:
@Fryguy This looks OK to me.
The `N_` notation means that strings would be marked for translation, but the actual translation would happen on the client side.

Reply:
Also change the Reason line to this -
`N_("Reason") => N_("VM does not exist")`

Reply:
I would also expect the hash to look something like `{:name => vm_name, :reason => "Some reason", ...}` and have the presenter handle translations.

Reply:
"Name" and "Reason" are presentation strings visible in the UI, yes.

But they are NOT translated in the model. 
By specifying `N_` we simply make sure that the i18n rake task will pick up those strings and put them in the gettext catalog, which will be then sent to the translators.